src/digraph.rs

```diff
@@ -939,6 +962,113 @@ impl PyDiGraph {
         Ok(index.index())
     }
 
+    /// Find node within this graph given a specific weight
+    ///
+    /// :param obj: The weight to look for
+    ///
+    /// returns: the index > 0 if a node was found and otherwise -1
+    /// rtype: int
+    pub fn find_node_by_weight(
+        &self,
+        py: Python,
+        obj: PyObject,
+    ) -> Option<usize> {
+        let mut index = None;
+        for node in self.graph.node_indices() {
+            let weight = self.graph.node_weight(node).unwrap();
+            let weight_compare = |a: &PyAny, b: &PyAny| -> PyResult<bool> {
+                let res = a.compare(b)?;
+                Ok(res == Ordering::Equal)
+            };
+
+            if weight_compare(obj.as_ref(py), weight.as_ref(py)).unwrap() {
+                index = Some(node.index());
+                break;
+            }
+        }
+        index
+    }
+
+    ///     Returns the graph that results from contracting u and v.
+    /// Node contraction identifies the two nodes as a single node incident to any edge that was incident to the original two nodes.
+    /// Parameters:
+    /// G (NetworkX graph) – The graph whose nodes will be contracted.
+    /// v (u,) – Must be nodes in G.
+    /// self_loops (Boolean) – If this is True, any edges joining u and v in G become self-loops on the new node in the returned graph.
+    /// Returns:
+    /// A new graph object of the same type as G (leaving G unmodified) with u and v identified in a single node. The right node v will be merged into the node u, so only u will appear in the returned graph.
+    /// Return type:
+    /// Networkx graph
+
+    /// Merge two nodes in the graph.
+    ///
+    /// :param node0: The source node that is going to be merged
+    /// :param node1: The target node that is going to be the new node
+    #[text_signature = "(u, v /)"]
+    pub fn merge_nodes(
+        &mut self,
+        py: Python,
+        u: usize,
+        v: usize,
+    ) -> PyResult<()> {
+        let source_node = NodeIndex::new(u);
+        let target_node = NodeIndex::new(v);
+
+        let source_weight = match self.graph.node_weight(source_node) {
+            Some(weight) => weight,
+            None => {
+                return Err(PyIndexError::new_err("No node found for index"))
+            }
+        };
+
+        let target_weight = match self.graph.node_weight(target_node) {
+            Some(weight) => weight,
+            None => {
+                return Err(PyIndexError::new_err("No node found for index"))
+            }
+        };
+
+        let compare_weights = |a: &PyAny, b: &PyAny| -> PyResult<bool> {
+            let res = a.compare(b)?;
+            Ok(res == Ordering::Equal)
+        };
+
+        let have_same_weights =
+            compare_weights(source_weight.as_ref(py), target_weight.as_ref(py));
```

Comment:
Small nit: I probably would eliminate the closure here so it's just something like:

```suggestion
        let have_same_weights =
             source_weight.as_ref(py).compare(target_weight.as_ref(py))?
                == Ordering::Equal;
```

just speeds up the compile a little. But not a big deal either way.

Reply:
If you do keep the closure you should add a `?` at the end of the call so that we return the inner exception raised by `compare()` to the user. This also eliminates the need for `unwrap()` below.

Reply:
fixed